crates/wasmparser/src/validator/operators.rs

```diff
@@ -1612,6 +1613,32 @@ where
         };
         Ok(())
     }
+    fn visit_global_atomic_get(
+        &mut self,
+        _ordering: crate::Ordering,
+        global_index: u32,
+    ) -> Self::Output {
+        // The following is almost identical to `visit_global_get` except for
+        // some additional type restrictions. No validation of `ordering` is
+        // needed because `global.atomic.get` can be used on both shared and
+        // unshared globals.
+        if let Some(ty) = self.resources.global_at(global_index) {
+            let ty = ty.content_type;
+            // TODO: need to express that any ValType::Ref(..) must be a subtype
+            // (<:) of `anyref`.
```

Comment:
IIRC there's various methods we have to get the "top type" and for `ValType::Ref` types you're interested in the ones where the "top type" is `Anyref`

Reply:
Stumbled across `self.resources.is_subtype(...)` which explains things well &mdash; what do you think of that?